src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1990,48 +2019,43 @@ Symbol* SystemDictionary::check_signature_loaders(Symbol* signature,
                                                        Symbol* signature,
                                                        TRAPS) {
+
+  MutexLocker ml(THREAD, InvokeMethodTable_lock);
   methodHandle empty;
   const int iid_as_int = vmIntrinsics::as_int(iid);
   assert(MethodHandles::is_signature_polymorphic(iid) &&
          MethodHandles::is_signature_polymorphic_intrinsic(iid) &&
          iid != vmIntrinsics::_invokeGeneric,
          "must be a known MH intrinsic iid=%d: %s", iid_as_int, vmIntrinsics::name_at(iid));
 
-  unsigned int hash  = invoke_method_table()->compute_hash(signature, iid_as_int);
-  int          index = invoke_method_table()->hash_to_index(hash);
-  SymbolPropertyEntry* spe = invoke_method_table()->find_entry(index, hash, signature, iid_as_int);
-  methodHandle m;
-  if (spe == NULL || spe->method() == NULL) {
-    spe = NULL;
-    // Must create lots of stuff here, but outside of the SystemDictionary lock.
-    m = Method::make_method_handle_intrinsic(iid, signature, CHECK_NULL);
-    if (!Arguments::is_interpreter_only() || iid == vmIntrinsics::_linkToNative) {
-      // Generate a compiled form of the MH intrinsic.
+  Method** met;
+  InvokeMethodKey key(signature, iid_as_int);
+  met = _invoke_method_intrinsic_table.get(key);
+
+  if (met != nullptr) {
+      return *met;
+  }
+
+  methodHandle m = Method::make_method_handle_intrinsic(iid, signature, CHECK_NULL);
+  if (!Arguments::is_interpreter_only() || iid == vmIntrinsics::_linkToNative) {
+      // Generate a compiled form of the MH intrinsic
       // linkToNative doesn't have interpreter-specific implementation, so always has to go through compiled version.
       AdapterHandlerLibrary::create_native_wrapper(m);
       // Check if have the compiled code.
       if (!m->has_compiled_code()) {
         THROW_MSG_NULL(vmSymbols::java_lang_VirtualMachineError(),
                        "Out of space in CodeCache for method handle intrinsic");
       }
-    }
-    // Now grab the lock.  We might have to throw away the new method,
-    // if a racing thread has managed to install one at the same time.
-    {
-      MutexLocker ml(THREAD, SystemDictionary_lock);
-      spe = invoke_method_table()->find_entry(index, hash, signature, iid_as_int);
-      if (spe == NULL)
-        spe = invoke_method_table()->add_entry(index, hash, signature, iid_as_int);
-      if (spe->method() == NULL)
-        spe->set_method(m());
-    }
   }
-
-  assert(spe != NULL && spe->method() != NULL, "");
-  assert(Arguments::is_interpreter_only() || (spe->method()->has_compiled_code() &&
-         spe->method()->code()->entry_point() == spe->method()->from_compiled_entry()),
+  // Now grab the lock.  We might have to throw away the new method,
+  // if a racing thread has managed to install one at the same time.
+  signature->make_permanent(); // The signature is never unloaded.
+  _invoke_method_intrinsic_table.put(key, m());
+  assert(m() != NULL, "");
```

Comment:
This assert seems odd. It seems to correspond to the old assert `spe->method() != NULL`, but that one says "we must have a hashtable entry that contains a non-null method". In the new code, we don't deal with hashtable entries directly anymore, so this assert can be removed.